README.md
# VIVA+ Open Human Body Models

Human body finite element (FE) models for Injury Assessments

*The VIVA+ models are under active development, with frequent beta releases. See updates in [CHANGELOG](CHANGELOG.md)*

## Quick start

Download the files (download [link above](https://openvt.eu/fem/viva/vivaplus/-/archive/master/vivaplus-master.zip))

or

Clone the repository using git:

```
git clone https://openvt.eu/fem/viva/vivaplus.git
```

### Directory structure

The include files that are common to all the models (material/section/property definitions, contacts, constraints, etc.) are placed in the `common` sub-directory. The model-specific files (main files, node coordinates) can be found in the respective sub-directory, e.g., `model/50F-Seated/`.

```
vivaplus
├─── model
│   ├─── 50F-seated
|   ├─── 50F-standing
│   ├─── 50F-seated
|   ├─── common
|   ├─── positioned_models
|   ├─── postproces
│   └─── preprocess
└─── docs
```

## Documentation and User Community

The documentation is available at <https://vivaplus.readthedocs.io/>

Questions? Join the conversation at **users' community** [![project chat](https://img.shields.io/badge/zulip-join_chat-brightgreen.svg)](https://vivaplus.zulipchat.com)

## How to cite

If you use the models or simulations, cite as

John, J., Klug, C., Kranjec, M., Svenning, E., & Iraeus, J., Frontiers in Bioengineering and Biotechnology (2022). Hello, World! VIVA+: A Human Body Model lineup to evaluate Sex-Differences in Crash Protection.

```
@Article{John2022,
  AUTHOR = {John, Jobin and Klug, Corina and Kranjec, Matej and Svenning, Erik and Iraeus, Johan},    
  TITLE = {Hello, world! VIVA+: A human body model lineup to evaluate sex-differences in crash protection},      
  JOURNAL = {Frontiers in Bioengineering and Biotechnology},       
  VOLUME = {10},           
  YEAR = {2022},        
  URL = {https://www.frontiersin.org/articles/10.3389/fbioe.2022.918904},       
  DOI = {10.3389/fbioe.2022.918904}, 
}
```

## License

The models are licensed under [GNU Lesser General Public License, v3](https://www.gnu.org/licenses/lgpl-3.0-standalone.html)

![LGPLv3)](docs/images/lgplv3.png)

&copy; 2019-2022, VIVA+ Developers (See `CONTRIBUTING.md` for details)
&copy; 2019, ViVA Developers (See `CONTRIBUTING.md` for details)

## Acknowledgement

![VIRTUAL Funding](docs/images/VIRTUAL_EUFunding.png)

See [projectvirtual.eu](https://projectvirtual.eu/) for more details.

CHANGELOG.md
# **Changelog**

All changes to this project are documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

# beta  versions

## 0.3.3 - 2022-09-08

### Modified

- Updates to Pelvic Floor for stability
  - material changed from `MAT_ELASTIC` to `MAT_FABRIC`
  - `ELFORM` changed from 16 to 9 on `SECID` 606004 and symmetry

## 0.3.2a - 2022-08-11

### Modified

- Temporary changes in pelvic floor and oral cavity for robustness

## 0.3.2 - 2022-05-16

### Added

- New output definitions for Mid-sternum and occipital condyles
- Positioned Cyclist models added on 2022-07-08

## 0.3.1 - 2022-04-29

### Added

- New PIDs (403603, 453603)at the junction between the costal cartilage and ribs plus at the junction of the posterior rigid ribs and the ribs (PIDs 403604, 453604) - to improve post processing of ribs

### Modified

- Updated contact 400001 to include also soft tissue around ribcage
  - Remaining part (abdominal muscle) of old contact 400010 was assigned to a new tied contact
  - Removed old contact 400010 and associated sets
  - Move nodes in soft tissue to work with new contact definitions

## 0.3.0rc4 - 2022-03-03

### Added

- Pedestrian Postures
- New global head accelerometer
- Missing left elbow landmark
- Recommended control cards in model director

### Modified

- Database extent binary card



## 0.3.0rc3 - 2022-03-03

This version is released on a new `git` repository. VIVA+ versions `0.2` can be accessed on the archived repo.

Bulk of the development for this version was done in ANSA and outside the repo, and hence the commits do not necessarily reflect the contributions.
The major contributions to this versions are:

- New upper extremity mesh definitions, and other mesh updates - Jobin
- Derivative models (seated male, standing male, standing female), abdominal muscles - Johan
- Updates to head and property tuning - Corina

### Added

- New density parameter for upper extremity, `UX_DENS`.
- New PID `100005` with deformable skull definition, with part of the skull assigned with the new PID.
- New Ulna and Radius mesh with solid definitions (previously shell)
- New hand definitions, with rigid wrist and deformable fingers
- New mesh pattern for foot, with rigid Tarsal/Metatarsal blob with soft tissue around it
- New PID for `UX-Bone-Radius-distal-Trabecular`: 301532, 351532
- Shoulder updates
  - Thorax muscles Rhomboideus Major, Trapezius Ascendens, Serratus Anterior, Subscapularis (23 muscles elements in each side)
    - The extra node sets constraining scapula and clavicle to soft tissues are removed with the addition of these muscles
  - Ligaments between scapula and clavicle, in the acromiclavicular joint (to prevent clavicle rotation around its own axis): `UX-Trapezoid-Ligament`, `UX-Conoid-Ligament`
  - Slightly updated position of scapula to remove penetrations with ribcage and skin after adding subscapularis
  - New contacts for scapula and clavicle `400002`
- Abdominal muscles PID `506002` connecting pelvis and ribcage
- Added nulls shells `PE-Pelvic-Sacrum-Cortical-Nulls` to cover sacrum holes
- Pelvic floor defined, PID `606004`
- New file for output definitions `vivaplus-90-outputs-nodes.k`, `vivaplus-elements.k`
  - Removed the old include `vivaplus-90-outputs.k`
- New include `vivaplus-92-output-settings.k` with output settings
  - Added Head_COG (seatbelt accelerometer)
- Added extra nodes connecting T4 down to L1 to the internal blob
- Added DAMPING_PART_STIFFNESS with 5% damping to all elastic materials
- New contact to tie soft tissue to humerus, 350010
- Names added `*DEFINE_COORDINATE_NODES`
- Added missing references for material

### Changed

- MIDs, Cordinate system IDs renumbered to be consistent with PID and body regions.
- Trimmed down and renumbered `*HOURGLASS` IDs
- Contact thickness `OPTT` in clavicle and scapula increased to 0.5
- Trimmed down Humerus null into single PID for each side
- UX-Bone-Humerus-proximal-Rigid is now part of null shells (was part of solids previously)
- UX-Soft-Inner-Surface-Null thickness changed to 0.5 (`UX-Soft-LowerArm-Null` removed and merged with this one)
- Updated contact thickness for rib cortical to constant `OPTT` = 0.5, instead varying nodal thickness
- Null PID `404001` (TX-Intercotal-muscle-External-T1-T12-L-Nulls) changed to `404401` to cover only innermost intercostal muscles
- Hip ligaments `603110`: changed `ELFORM` from 9 to 16 to add bending stiffness, changed material definition and add `OPTT` = 1
- Moved proximal tibia cortical elements to new PID `701411`
- Add `ICOMP` = 1 to PID `705191`
- Head
  - Skull is combination deformable and rigid areas
  - Scalp properties changed from Hybrid-III rubber to fat tissue (Naseri)
  - Eyelids changed to skin properties and a layer of solid elements added behind the eyelids for stabilizing the membrane
- Neck
  - Trimmed down MIDs. All vertebral levels refer to same MIDs for cortical, trabecular, endplate material definitions.
  - Tied contacts in the neck have `IGNORE` = 1 
- New force outputs defined for thoracic spine
- Contacts
  - Aligned friction coefficient for all contacts to 0.05
  - Changed to `SOFT`=1, `SBOPT` to zero and `DEPTH` to 2 for contacts
    - `100001`
    - `200002`
    - `202100`
    - `300000` (UX_Single_Surface)
    - `300001`
    - `400001` (depenetrated based on new penetrations reported by ANSA)
  - Changed `300001` from `*CONTACT_AUTOMATIC_SURFACE_TO_SURFACE` to `*CONTACT_AUTOMATIC_SINGLE_SURFACE`
  - Soft Tissue around humerus is now tied with `TIED_NODES_TO_SURFACE_OFFSET` (`300010`, `350010`)
  - 
  - `400010` Between Torso soft tissue and ribcage updated to `TIED_SHELL_EDGE_TO_SURFACE_BEAM_OFFSET` from `AUTOMATIC_SURFACE_TO_SURFACE_TIEBREAK`
    - Deleted segments previously used for this contact
  - `403505` Master updated to shell set that has soft tissue shells against sternum only

utils/ansa/heading_name_script.py
# PYTHON script for ANSA
# This script renames the *DATABASE_HISTORY_NODE Name based on the Heading of the Node it references
import os
import ansa
from ansa import base, constants

def main():
	
	history = base.CollectEntities(constants.LSDYNA, None, "DATABASE_HISTORY_NODE", recursive=True) # Collect all DATABASE_HISTORY_NODE
	for h in history:
		vals = ('No.of.Nodes', )
		number_of_nodes = base.GetEntityCardValues(entity=h, fields=vals, deck=constants.LSDYNA)['No.of.Nodes']   # Get number of nodes in the DATABASE_HISTORY_NODE list
		for n in range(1,int(number_of_nodes)):    # Loop the list of nodes
			n_id = 'NODE'+str(n)
			heading_id =  'Name'
			vals = (n_id, )
			node_id = base.GetEntityCardValues(entity=h, fields=vals, deck=constants.LSDYNA) # Get Node id
			node = base.GetEntity(ansa.constants.LSDYNA, "NODE", node_id[n_id]) # Get Node entity
			vals = ('Name',)
			node_name = base.GetEntityCardValues(entity=node, fields=vals, deck=constants.LSDYNA)['Name'] # Get Node Name
			
			fields = {heading_id: node_name,}
			base.SetEntityCardValues(constants.LSDYNA, h, fields) # Set the new HEADING name = node name

if __name__ == '__main__':
	main()




assets/part-numbering/PID-assign-script.py
# PYTHON script
import os
import ansa
import csv
from ansa import *

def main():
	path0="C:/viva-plus/assets/part-numbering/"
	
	body_region = 'NE'
	
	if body_region == 'HE':
		body_region_dir = '100000-HE'
	elif body_region == 'NE':
		body_region_dir = '200000-neck'
	elif body_region == 'UX':
		body_region_dir = '300000-UX'
	elif body_region == 'TX':
		body_region_dir = '400000-TX'
	elif body_region == 'AB':
		body_region_dir = '500000-AB'
	elif body_region == 'PE':
		body_region_dir = '600000-PE'
	elif body_region == 'LX':
		body_region_dir = '700000-LX'
	else :
		print('No Body region given')
        
	#body_region_dir="200000-neck/"
	file="202100-neck-ivd-annulus-fibers"
	with open(path0+body_region_dir+"/"+file+"-PID-PIDName.csv", mode='r') as csv_file:
	#with open('C:/viva-plus/documentation/assets/part-numbering/600000-Pelvis/601000-Pelvis-Bones-PID.csv', mode='r') as csv_file: # read the csv file with the format pid_id,pid_name
		csv_reader = csv.reader(csv_file, delimiter=',')
		pid_dict = dict()
		for row in csv_reader:						# Create dictionary with pid_id:pid_name
			pid_dict[row[0]] = row[1]
# Loop all PIDs  and assign new name according to the dictionaries
	for entity in base.CollectEntitiesI(constants.LSDYNA, None, '__PROPERTIES__'):
		if str(entity._id) in pid_dict:	#check if PID is avaiable in the dict/csv
			vals = {'Name': pid_dict[str(entity._id)],}
			base.SetEntityCardValues(constants.LSDYNA, entity, vals)

# Loop all MIDs and assign new name according to the dictionaries (if already defined in ANSA)
	for entity in base.CollectEntitiesI(constants.LSDYNA, None, '__MATERIALS__'):
		if str(entity._id) in pid_dict:	#check if PID is avaiable in the dict/csv
			vals = {'Name': pid_dict[str(entity._id)],}
			base.SetEntityCardValues(constants.LSDYNA, entity, vals)

# For MID
#	with open('C:/ANSA/mid.csv', mode='r') as csv_file: # read the csv file with the format mid_id,mid_name
#		csv_reader = csv.reader(csv_file, delimiter=',')
#		mid_dict = dict()
#		for row in csv_reader:						# Create dictionary with mid_id:mid_name
#			mid_dict[row[0]] = row[1]
# Loop all MIDs and assign new name according to the dictionaries
#	for entity in base.CollectEntitiesI(constants.LSDYNA, None, '__MATERIALS__'):
#		vals = {'Name': mid_dict[str(entity._id)],}
#		base.SetEntityCardValues(constants.LSDYNA, entity, vals)
#

if __name__ == '__main__':
	main()


assets/part-numbering/part-naming-from-pid.py
#!/usr/bin/env python
# coding: utf-8

import numpy as np
import pandas as pd
import math
import os
import click

cwd = os.getcwd()

# Command Line Interface using click

@click.command()
@click.option('--body_region', prompt='Which body region?', default=' ', help='Body Region: HE, NE, UX, TX, AB, PE, LX')
@click.option('--part_group', prompt='Part group CSV', default=' ',
              help='Name of the CSV containing the ID naming scheme')


def PID_name_from_scheme_main(body_region, part_group):
    ''' Compute PID name based on the PID numbering scheme given in CSV format
        
        Input: CSV file with PID Numbering Scheme
        
        Output: CSV in the format PID-PID names, which serves as the input for ANSA PID-assign-script
        
        '''

    if body_region == 'HE':
        body_region_dir = '100000-HE'
    elif body_region == 'NE' :
        body_region_dir = '200000-NE'
    elif body_region == 'UX' :
        body_region_dir = '300000-UX'
    elif body_region == 'TX' :
        body_region_dir = '400000-TX'
    elif body_region == 'AB' :
        body_region_dir = '500000-AB'
    elif body_region == 'PE' :
        body_region_dir = '600000-PE'
    elif body_region == 'LX' :
        body_region_dir = '700000-LX'
    else :
        print('No Body region given')
        

    if part_group == ' ':
        print('No Part Group given')

    csv_file = os.path.join(cwd, body_region_dir, part_group)


    # Read the csv file
    data = pd.read_csv(csv_file+".csv")
    #data = pd.read_csv("../part-numbering/600000-Pelvis/601000-Pelvis-Bones.csv") 

    PID_names = assign_name(data)
    output_to_csv(PID_names, csv_file)


    pid = 0
    name = ''

    # FIXME: Error while opening some csv files
    # the type of for loop iter variable for PID (e.g. fi1) should be int,
    # sometimes ends up as string (for some CSV files. 
    # The PID values are not computed then.
    # include type conversion?


def assign_name(data):
    '''Assign PID name based on digits in the PID numbering scheme

    Arguments: dataframe from the PID naming scheme

    Returns a PID-PIDName dictionary 
    '''
    dic = {}
    num_of_rows=data.shape[0]

# Extract the digits present in the csv (now in dataframe) and assign to a list corresponding to its place

    for n in range(num_of_rows):
            if pd.notna(data.iloc[n,0]):
                first = data.iloc[n,0]
                body_region = str(data.iloc[n,6])
                
                second = []
                third = []
                fourth = []
                fifth = []
                sixth = []
                
                sagittal_aspect = []
                third_name = []
                fourth_name = []
                fifth_name = []
                sixth_name = []
                continue
                
            if pd.notna(data.iloc[n,1]):
                second.append(int(data.iloc[n,1]))
                sagittal_aspect.append(str(data.iloc[n,6]))
                #print(second,sagittal_aspect)
                continue
                
            if pd.notna(data.iloc[n,2]):
                third.append(int(data.iloc[n,2]))
                #if pd.notna(data.iloc[n,6]):
                third_name.append(str(data.iloc[n,6]))
                continue
                
            if pd.notna(data.iloc[n,3]):
                fourth.append(int(data.iloc[n,3]))
                #if pd.notna(data.iloc[n,6]):
                fourth_name.append(str(data.iloc[n,6]))
                continue
                
            if pd.notna(data.iloc[n,4]):
                fifth.append(int(data.iloc[n,4]))
                #if pd.notna(data.iloc[n,6]):
                fifth_name.append(str(data.iloc[n,6]))
                continue
                
            if pd.notna(data.iloc[n,5]):
                sixth.append(int(data.iloc[n,5]))
                sixth_name.append(str(data.iloc[n,6]))
                #print(sixth,sixth_name)
                
    # Loop through the list to create PID and PID names
    # Takes the digits and corresponding names in the csv and writes them to a dictionary

    if not first:
        print("Error processing CSV. Seems to empty dataframe")

    for s1,s2 in zip(second,sagittal_aspect):
        for t1,t2 in zip(third,third_name):
            if not fifth:
                if fourth == []:
                    
                    # Check if the sixth place in the csv is empty
                    if sixth == []:
                        
                        # print(s1,t1,fi1,s2,t2,fi2)
                                    #if math.isnan(si1):
                                        #si1=0
                    
                                    #if math.isnan(f1):
                        pid = int(first*10**5 + s1*10**4 + t1*10**3)    
                        name = body_region + '-' + str(t2) + '-'  + str(s2)
                        dic[pid]= name
                                    #else:
                                    # pid = int(first*10**5 + s1*10**4 + t1*10**3 + f1*100 +si1)    
                                        #name = body_region + '-' + str(t2) + '-' + str(f2)+ '-' + str(si2) + '-' + str(s2)
                                    #dic[pid]= name
                                        
                    # Actions if sixth place is not empty
                    else:
                        for si1,si2 in zip(sixth,sixth_name):
                            # Test print
                            # print(s1,t1,fi1,si1,s2,t2,fi2,si2)
                            pid = int(first*10**5 + s1*10**4 + t1*10**3 + si1)    
                            name = body_region + '-' + str(t2) + '-' + str(si2) + '-' + str(s2)
                            dic[pid]= name

                # Actions if the fourth place is not empty            
                else:    
                    for f1,f2 in zip(fourth,fourth_name):
                        # Check if the sixth place is empty
                        if sixth == []:
                        
                            #print(s1,t1,f1,fi1,s2,t2,f2,fi2)
                                    #if math.isnan(si1):
                                        #si1=0
                    
                                    #if math.isnan(f1):
                            pid = int(first*10**5 + s1*10**4 + t1*10**3 + f1*100)    
                            name = body_region + '-' + str(t2) + '-' + str(f2) + '-'  + str(s2)
                            dic[pid]= name
                                    #else:
                                    # pid = int(first*10**5 + s1*10**4 + t1*10**3 + f1*100 +si1)    
                                        #name = body_region + '-' + str(t2) + '-' + str(f2)+ '-' + str(si2) + '-' + str(s2)
                                    #dic[pid]= name
                        else:
                            for si1,si2 in zip(sixth,sixth_name):
                                # Test print
                                # print(s1,t1,f1,fi1,si1,s2,t2,f2,fi2,si2)
                                pid = int(first*10**5 + s1*10**4 + t1*10**3 + f1*100 +si1)    
                                name = body_region + '-' + str(t2) + '-' + str(f2) + '-' + str(si2) + '-' + str(s2)
                                dic[pid]= name
            else:
                for fi1,fi2 in zip(fifth,fifth_name):
                    # Check if the fourth place in the csv is empty
                    #print(fourth)
                    if fourth == []:
                        
                        # Check if the sixth place in the csv is empty
                        if sixth == []:
                            
                            # print(s1,t1,fi1,s2,t2,fi2)
                                        #if math.isnan(si1):
                                            #si1=0
                        
                                        #if math.isnan(f1):
                            pid = int(first*10**5 + s1*10**4 + t1*10**3 + fi1*10)    
                            name = body_region + '-' + str(t2) + '-' + str(fi2) + '-'  + str(s2)
                            dic[pid]= name
                                        #else:
                                        # pid = int(first*10**5 + s1*10**4 + t1*10**3 + f1*100 + fi1*10 +si1)    
                                            #name = body_region + '-' + str(t2) + '-' + str(f2) + '-'+ str(fi2) + '-' + str(si2) + '-' + str(s2)
                                        #dic[pid]= name
                                            
                        # Actions if sixth place is not empty
                        else:
                            for si1,si2 in zip(sixth,sixth_name):
                                # Test print
                                # print(s1,t1,fi1,si1,s2,t2,fi2,si2)
                                pid = int(first*10**5 + s1*10**4 + t1*10**3 + fi1*10 + si1)    
                                name = body_region + '-' + str(t2) + '-' + str(fi2) + '-' + str(si2) + '-' + str(s2)
                                dic[pid]= name

                    # Actions if the fourth place is not empty            
                    else:    
                        for f1,f2 in zip(fourth,fourth_name):
                            # Check if the sixth place is empty
                            if sixth == []:
                            
                                #print(s1,t1,f1,fi1,s2,t2,f2,fi2)
                                        #if math.isnan(si1):
                                            #si1=0
                        
                                        #if math.isnan(f1):
                                pid = int(first*10**5 + s1*10**4 + t1*10**3 + f1*100 + fi1*10)    
                                name = body_region + '-' + str(t2) + '-' + str(f2) + '-' + str(fi2) + '-'  + str(s2)
                                dic[pid]= name
                                        #else:
                                        # pid = int(first*10**5 + s1*10**4 + t1*10**3 + f1*100 + fi1*10 +si1)    
                                            #name = body_region + '-' + str(t2) + '-' + str(f2) + '-'+ str(fi2) + '-' + str(si2) + '-' + str(s2)
                                        #dic[pid]= name
                            else:
                                for si1,si2 in zip(sixth,sixth_name):
                                    # Test print
                                    # print(s1,t1,f1,fi1,si1,s2,t2,f2,fi2,si2)
                                    pid = int(first*10**5 + s1*10**4 + t1*10**3 + f1*100 + fi1*10 +si1)    
                                    name = body_region + '-' + str(t2) + '-' + str(f2) + '-' + str(fi2) + '-' + str(si2) + '-' + str(s2)
                                    dic[pid]= name
    
    print("PID - PIDNames generated")
    return dic

                                        
    # print (dic)


    # In[32]:

def output_to_csv(dic, csv_file):
    '''Output PID - PIDName dictionary as csv'''

    # Output the dictionary values to a csv
    (pd.DataFrame.from_dict(data=dic, orient='index')
        .to_csv(csv_file+"-PID-PIDName" + ".csv", header=False))
    print("PID-PIDName file output: \n"+csv_file+"-PID-PIDName" + ".csv")

if __name__ == '__main__':
    PID_name_from_scheme_main()

utils/part-numbering/ANSA-PID-assign-script.py
# PYTHON script
import os
import ansa
import csv
from ansa import *

def main():
	path0="C:/openvt/vivaplus/utils/part-numbering/"
	
	body_region = 'TX'
		#body_region_dir="200000-neck/"
	#file="202100-neck-ivd-annulus-fibers"
	file="404000-TX-Intercostal-Muscle"
	
	if body_region == 'HE':
		body_region_dir = '100000-HE'
	elif body_region == 'NE':
		body_region_dir = '200000-NE'
	elif body_region == 'UX':
		body_region_dir = '300000-UX'
	elif body_region == 'TX':
		body_region_dir = '400000-TX'
	elif body_region == 'AB':
		body_region_dir = '500000-AB'
	elif body_region == 'PE':
		body_region_dir = '600000-PE'
	elif body_region == 'LX':
		body_region_dir = '700000-LX'
	else :
		print('No Body region given')
        

	with open(path0+body_region_dir+"/"+file+"-PID-PIDName.csv", mode='r') as csv_file:
	#with open('C:/viva-plus/documentation/assets/part-numbering/600000-Pelvis/601000-Pelvis-Bones-PID.csv', mode='r') as csv_file: # read the csv file with the format pid_id,pid_name
		csv_reader = csv.reader(csv_file, delimiter=',')
		pid_dict = dict()
		for row in csv_reader:						# Create dictionary with pid_id:pid_name
			pid_dict[row[0]] = row[1]
# Loop all PIDs  and assign new name according to the dictionaries
	for entity in base.CollectEntitiesI(constants.LSDYNA, None, '__PROPERTIES__'):
		if str(entity._id) in pid_dict:	#check if PID is avaiable in the dict/csv
			vals = {'Name': pid_dict[str(entity._id)],}
			base.SetEntityCardValues(constants.LSDYNA, entity, vals)

# Loop all MIDs and assign new name according to the dictionaries (if already defined in ANSA)
	for entity in base.CollectEntitiesI(constants.LSDYNA, None, '__MATERIALS__'):
		if str(entity._id) in pid_dict:	#check if PID is avaiable in the dict/csv
			vals = {'Name': pid_dict[str(entity._id)],}
			base.SetEntityCardValues(constants.LSDYNA, entity, vals)

# For MID
#	with open('C:/ANSA/mid.csv', mode='r') as csv_file: # read the csv file with the format mid_id,mid_name
#		csv_reader = csv.reader(csv_file, delimiter=',')
#		mid_dict = dict()
#		for row in csv_reader:						# Create dictionary with mid_id:mid_name
#			mid_dict[row[0]] = row[1]
# Loop all MIDs and assign new name according to the dictionaries
#	for entity in base.CollectEntitiesI(constants.LSDYNA, None, '__MATERIALS__'):
#		vals = {'Name': mid_dict[str(entity._id)],}
#		base.SetEntityCardValues(constants.LSDYNA, entity, vals)
#

if __name__ == '__main__':
	main()
